consensus/consensus.go

```diff
@@ -1,103 +1,237 @@
+	gb *blockchain.Block) (bool, GenesisBlockCode) {
+
+	// Check if the genesis block is equal to nil.
+	if gb == nil {
+		return false, NilGenesisBlock
+	}
 
-	// Check if the block is valid
-	if valid, code := bc.ValidBlock(b); !valid {
-		log.Errorf("Invalid block, BlockCode: %d", code)
-		return false, BadBlock
+	// Check if the genesis block's block number is equal to 0.
+	if gb.BlockHeader.BlockNumber != 0 ||
+		bc.Blocks[0] != gb {
+		return false, BadGenesisBlockNumber
 	}
 
-	// Check if the CloudBase transaction reward is equal to the network's
-	// current block reward
-	if b.GetCloudBaseTransaction().Outputs[0].Amount != BlockReward {
-		log.Error("Invalid Block Reward")
-		return false, BadBlockReward
+	// Check if the genesis block's last block hash is equal to 0.
+	if blockchain.HashToBigInt(gb.BlockHeader.LastBlock).Cmp(c.Big0) != 0 {
+		return false, BadGenesisLastBlock
 	}
 
-	// Check if the CloudBase transaction recipient is equal to the address
-	// of the miner that mined the block
-	if b.GetCloudBaseTransaction().Outputs[0].Recipient != cb {
-		log.Error("Invalid CloudBase address")
-		return false, BadCloudBase
+	// Check if the size of the transaction list is equal to 1.
+	if len(gb.Transactions) != 1 {
+		return false, BadGenesisTransactions
 	}
 
-	// Check if the block's target is equal to the network's current target
-	if b.Target != CurrentTarget() {
-		log.Error("Invalid Target")
+	// Check if the transaction is a valid cloud base transaction.
+	if valid, code := VerifyCloudBase(bc, gb.Transactions[0]); !valid {
+		log.Errorf("Invalid CloudBase, CloudBaseTransactionCode: %d", code)
+		return false, BadGenesisCloudBaseTransaction
+	}
+
+	// Check that the target is within the min and max difficulty levels.
+	// TODO: CurrentTarget() is used because the difficulty is static for v1 of
+	// cumulus. Once the difficulty is dynamic, the target would need to be
+	// compared to the target calculated for that specific block.
+	target := blockchain.HashToBigInt(gb.Target)
+	if target.Cmp(c.MaxTarget) == 1 ||
+		target.Cmp(c.MinTarget) == -1 ||
+		target.Cmp(blockchain.HashToBigInt(CurrentTarget())) != 0 {
+		return false, BadGenesisTarget
+	}
+
+	// Check that time is not greater than current time or equal to 0.
+	if uint32(gb.Time) == 0 {
+		return false, BadGenesisTime
+	}
+
+	return true, ValidGenesisBlock
+}
+
+// VerifyBlock checks whether a block is valid.
+func VerifyBlock(bc *blockchain.BlockChain,
+	b *blockchain.Block) (bool, BlockCode) {
+
+	// Check if the block is equal to nil.
+	if b == nil {
+		return false, NilBlock
+	}
+
+	// Check if the block is the genesis block.
+	if b.BlockHeader.BlockNumber == 0 || bc.Blocks[0] == b {
+		if valid, code := VerifyGenesisBlock(bc, b); !valid {
+			log.Errorf("Invalid GenesisBlock, GenesisBlockCode: %d", code)
+			return false, BadGenesisBlock
+		}
+		return true, ValidBlock
+	}
+
+	// Check that block number is between 0 and max blocks.
```

Comment:
I would change this to: "Check that block number is valid" or something similar because "max blocks" isn't really correct terminology here.